intervals/main_objs.py

```diff
@@ -1417,66 +1565,48 @@ def homorhythm(self):
+        # active version with lyric ngs
         nr = self.notes()
         dur = self.durations(df=nr)
         ng = self.ngrams(df=dur, n=2)
         dur_ngrams = []
-
-        # find passages with more than 2 active voices
         for index, rows in ng.iterrows():
-
-            dur_ngrams_no_nan = [x for x in rows if pd.isnull(x) == False]
-            dur_ngrams.append(dur_ngrams_no_nan)
+             dur_ngrams_no_nan = [x for x in rows if pd.isnull(x) == False]
+             dur_ngrams.append(dur_ngrams_no_nan)
 
         ng['dur_ngrams'] = dur_ngrams
-        # ng['rest_count'] = rests
+
         ng['active_voices'] = ng['dur_ngrams'].apply(len)
         ng['number_dur_ngrams'] = ng['dur_ngrams'].apply(set).apply(len)
-        ng = ng[(ng['number_dur_ngrams'] <2) & (ng['active_voices'] > 2)]
-
-        # check rests in multiple parts
-        nr.ffill(inplace=True)
-        index_of_rests = []
-        rests = []
-        for index, rows in nr.iterrows():
-            rest_test = [y for y in rows if y == "Rest"]
-            rests.append(rest_test)
-
-        #     index_of_rests.append(index)
-        nr["rests"] = rests
-        nr["rests_count"] = nr["rests"].apply(len)
-        full_stop = nr[(nr['rests_count'] > 1) ]
-        rests_with_mb = self.detailIndex(full_stop)
-        # now get lyric syllables
+        ng = ng[(ng['number_dur_ngrams'] < 2) & (ng['active_voices'] > 1)]
+
+         # get the lyrics as ngrams to match the durations
         lyrics = self.lyrics()
-        lyrics = lyrics.applymap(ImportedPiece._alpha_only)
-        cols = lyrics.columns
-        for col in cols:
-            lyrics[col] = lyrics[col].str.lower()
+        lyrics = lyrics.applymap(self._alpha_only)
+        lyrics_ng = self.ngrams(df=lyrics, n=2)
+
+        # count the lyric_ngrams at each position
         syll_set = []
-        for index2, rows2 in lyrics.iterrows():
-            syll_no_nan = [z for z in rows2 if pd.isnull(z) == False]
-            syll_set.append(syll_no_nan)
-        #     print(syll_no_nan)
-        lyrics['syllable_set'] = syll_set
-
-        # create mask consisting of passages with more than two voices actively singing same syllables
-        lyrics['active_syll_voices'] = lyrics['syllable_set'].apply(len)
-        # count how _many_ syllables at this offset
-        lyrics['number_sylls'] = lyrics['syllable_set'].apply(set).apply(len)
-        # get count of possible hr passages (several voices with same syllable)
-        lyrics_hr = lyrics[(lyrics['active_syll_voices'] > 2) & (lyrics['number_sylls'] < 2)]
-        # self.detailIndex(lyrics_hr, offset=True)
-        # lyrics['is_hr'] = np.where(lyrics['active_voices'] > 3)
-        hr_sylls_mask = lyrics_hr["active_syll_voices"]
-
-        # combine results to show passages where more than 2 voices have the same syllables and durations
+        for index, rows in lyrics_ng.iterrows():
+             syll_no_nan = [z for z in rows if pd.isnull(z) == False]
```

Comment:
In general if you want a list of non-nan values of a series (row), you could do:
rows.dropna().to_list()

Reply:
I will look into that